File: Assets/src/model/enemies/Blob.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;

[Serializable]
[ObjectInfo(description: "Telegraphs attacks for 1 turn.\nChases you.", flavorText: "An odorless mass advances towards you with a steady, brainless determination.")]
public class Blob : AIActor {
  public static Item HomeItem => new ItemGloopShoes();
  public override float turnPriority => task is AttackGroundTask ? 90 : base.turnPriority;
  public Blob(Vector2Int pos) : base(pos) {
    hp = baseMaxHp = 4;
    faction = Faction.Enemy;
  }

  internal override (int, int) BaseAttackDamage() {
    return (2, 2);
  }

  protected override ActorTask GetNextTask() {
    if (CanTargetPlayer()) {
      if (IsDiagonallyNextTo(GameModel.main.player)) {
        return new AttackGroundTask(this, GameModel.main.player.pos, 1);
      } else {
        var chase = new ChaseTargetTask(this, GameModel.main.player);
        chase.maxMoves = 1;
        return chase;
        // return new ChaseTargetTask(this, GameModel.main.player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }
}

File: web/src/model/enemies/Blob.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackGroundTask } from '../tasks/AttackGroundTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';

/**
 * Telegraphs attacks for 1 turn. Chases you.
 * Port of Blob.cs.
 */
export class Blob extends AIActor {
  get turnPriority(): number {
    return this.task?.constructor.name === 'AttackGroundTask' ? 90 : super.turnPriority;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 6;
  }

  baseAttackDamage(): [number, number] {
    return [2, 3];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackGroundTask(this, player.pos, 1);
      } else {
        return new ChaseTargetTask(this, player);
      }
    }
    return new MoveRandomlyTask(this);
  }
}

/**
 * Smaller version of Blob. Same behavior, less HP/damage.
 */
export class MiniBlob extends AIActor {
  get turnPriority(): number {
    return this.task?.constructor.name === 'AttackGroundTask' ? 90 : super.turnPriority;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 3;
  }

  baseAttackDamage(): [number, number] {
    return [2, 2];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackGroundTask(this, player.pos, 1);
      } else {
        return new ChaseTargetTask(this, player);
      }
    }
    return new MoveRandomlyTask(this);
  }
}
